assets/js/datastore.coffee

```diff
@@ -4,180 +4,123 @@ if module?
-      throw 'Not implemented'
-    getParent: (row) ->
-      throw 'Not implemented'
-    setParent: (row, parent) ->
-      throw 'Not implemented'
-    getChildren: (row) ->
-      throw 'Not implemented'
-    setChildren: (row, children) ->
-      throw 'Not implemented'
-
-    getCollapsed: (row) ->
-      throw 'Not implemented'
-    setCollapsed: (row, collapsed) ->
-      throw 'Not implemented'
-
-    # get global settings (data not specific to a document)
-    getSetting: (setting) ->
-      throw 'Not implemented'
-    setSetting: (setting, value) ->
-      throw 'Not implemented'
-
-    # get mapping of row -> mark, for subtree beneath row
-    getMarks: (row) ->
-      throw 'Not implemented'
-    # set mapping of row -> mark, for subtree beneath row
-    setMarks: (row, marks) ->
-      throw 'Not implemented'
-
-    # maintain global marks datastructure.  maps mark -> row
-    getAllMarks: () ->
-      throw 'Not implemented'
-    setAllMarks: (marks) ->
-      throw 'Not implemented'
+    constructor: (prefix='') ->
+      @prefix = "#{prefix}save"
 
-    # get next row ID
-    getId: () ->
-      throw 'Not implemented'
-    getNew: () ->
-      id = do @getId
-      @setLine id, []
-      @setChildren id, []
-      return id
+      @_lineKey_ = (row) -> "#{@prefix}:#{row}:line"
+      @_parentKey_ = (row) -> "#{@prefix}:#{row}:parent"
+      @_childrenKey_ = (row) -> "#{@prefix}:#{row}:children"
+      @_collapsedKey_ = (row) -> "#{@prefix}:#{row}:collapsed"
+      @_marksKey_ = (row) -> "#{@prefix}:#{row}:marks"
 
-    # get last view (for page reload)
-    setLastViewRoot: (row) ->
-      throw 'Not implemented'
-    getLastViewRoot: () ->
-      throw 'Not implemented'
+      # no prefix, meaning it's global
+      @_settingKey_ = (setting) -> "settings:#{setting}"
 
-    # delete: (id) ->
-    #  throw 'Not implemented'
+      @_lastSaveKey_ = "#{@prefix}:lastSave"
+      @_lastViewrootKey_ = "#{@prefix}:lastviewroot"
+      @_allMarksKey_ = "#{@prefix}:allMarks"
+      @_IDKey_ = "#{@prefix}:lastID"
 
-  class InMemory extends DataStore
-    constructor: () ->
-      @lines = {}
-      @parents = {}
-      @children = {}
-      @collapsed = {}
-      @marks = {}
+    get: (key, default_value=null) ->
+        throw 'Not implemented'
 
-      @settings = {}
-      @allMarks = {}
-      return
+    set: (key, value) ->
+        throw 'Not implemented'
 
+    # get and set values for a given row
     getLine: (row) ->
-      return [].slice.apply @lines[row]
-
+      [].slice.apply @get (@_lineKey_ row)
     setLine: (row, line) ->
-      @lines[row] = line
+      @set (@_lineKey_ row), line
 
     getParent: (row) ->
-      return @parents[row]
-
+      @get (@_parentKey_ row)
     setParent: (row, parent) ->
-      @parents[row] = parent
+      @set (@_parentKey_ row), parent
 
     getChildren: (row) ->
-      return [].slice.apply @children[row]
-
+      [].slice.apply @get (@_childrenKey_ row)
```

Comment:
getLine and getChildren needed defaults of [] for me - i think just cause my document had old stuff and it wasn't always populated before (tests passed though)